datafusion/expr/src/signature.rs

```diff
@@ -123,35 +126,211 @@ pub enum TypeSignature {
+            is_append: bool,
+            allow_null_coercion: bool,
+        ) -> Result<Vec<Vec<DataType>>> {
+            if current_types.len() != 2 {
+                return Ok(vec![vec![]]);
+            }
+
+            let (array_type, elem_type) = if is_append {
+                (&current_types[0], &current_types[1])
+            } else {
+                (&current_types[1], &current_types[0])
+            };
+
+            // We follow Postgres on `array_append(Null, T)`, which is not valid.
+            if array_type.eq(&DataType::Null) {
+                if allow_null_coercion {
+                    return Ok(vec![vec![array_type.clone(), elem_type.clone()]]);
+                } else {
+                    return Ok(vec![vec![]]);
+                }
+            }
+
+            // We need to find the coerced base type, mainly for cases like:
+            // `array_append(List(null), i64)` -> `List(i64)`
+            let array_base_type = datafusion_common::utils::base_type(array_type);
+            let elem_base_type = datafusion_common::utils::base_type(elem_type);
+            let new_base_type = comparison_coercion(&array_base_type, &elem_base_type);
+
+            let new_base_type = new_base_type.ok_or_else(|| {
+                internal_datafusion_err!(
+                    "Coercion from {array_base_type:?} to {elem_base_type:?} not supported."
+                )
+            })?;
+
+            let new_array_type =
+                datafusion_common::utils::coerced_type_with_base_type_only(
+                    array_type,
+                    &new_base_type,
+                );
+
+            match new_array_type {
+                DataType::List(ref field)
+                | DataType::LargeList(ref field)
+                | DataType::FixedSizeList(ref field, _) => {
+                    let new_elem_type = field.data_type();
+                    if is_append {
+                        Ok(vec![vec![new_array_type.clone(), new_elem_type.clone()]])
+                    } else {
+                        Ok(vec![vec![new_elem_type.to_owned(), new_array_type.clone()]])
+                    }
+                }
+                _ => Ok(vec![vec![]]),
+            }
+        }
+        fn array_and_index(
+            current_types: &[DataType],
+            allow_null_coercion: bool,
+        ) -> Result<Vec<Vec<DataType>>> {
+            if current_types.len() != 2 {
+                return Ok(vec![vec![]]);
+            }
+
+            let array_type = &current_types[0];
+
+            if array_type.eq(&DataType::Null) && !allow_null_coercion {
+                return Ok(vec![vec![]]);
+            }
+
+            match array_type {
+                DataType::List(_)
+                | DataType::LargeList(_)
+                | DataType::FixedSizeList(_, _) => {
+                    let array_type = coerced_fixed_size_list_to_list(array_type);
+                    Ok(vec![vec![array_type, DataType::Int64]])
```

Comment:
I still think it can be further simplied if there is a function that process `coerced_fixed_size_list_to_list` and return `array_type`.  Then can be used both in `array_and_index` and `array`